protocols/primary-backup/src/main/java/io/atomix/protocols/backup/roles/PrimaryRole.java

```diff
@@ -171,6 +170,9 @@ private ExecuteResponse applyQuery(ExecuteRequest request, Session session) {
     context.service().backup(buffer);
     buffer.flip();
     byte[] bytes = buffer.readBytes(buffer.remaining());
+
+    replicator.removePreviousOperation(request.nodeId(), context.currentIndex());
```

Comment:
This line seems to assume that the primary's response to the backup is going to be received. That seems like it could be dangerous to me. We probably shouldn't remove operations from the primary until it *knows* the backup received and installed the snapshot. Of course, we could just not remove these operations here at all. Instead, the backup can respond to backup requests with the highest index it has received, and the primary can remove indexes it _knows_ the backup already has.